File: cpp/core/fixel/correspondence/algorithms/legacy.cpp
/* Copyright (c) 2008-2026 the MRtrix3 contributors.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Covered Software is provided under this License on an "as is"
 * basis, without warranty of any kind, either expressed, implied, or
 * statutory, including, without limitation, warranties that the
 * Covered Software is free of defects, merchantable, fit for a
 * particular purpose or non-infringing.
 * See the Mozilla Public License v. 2.0 for more details.
 *
 * For more details, see http://www.mrtrix.org/.
 */

#include "fixel/correspondence/algorithms/legacy.h"

#include "app.h"

namespace MR::Fixel::Correspondence::Algorithms {

// clang-format off
App::OptionGroup LegacyOptions = App::OptionGroup("Options specific to algorithm \"legacy\"")
  + App::Option("angle",
           "maximum angle within which a corresponding fixel may be selected, in degrees"
           " (default: " + str(default_nearest_maxangle) + ")")
    + App::Argument("value").type_float(0.0f, 90.0f);
// clang-format on

std::vector<std::vector<Mapping::Entry>> Legacy::operator()(const voxel_t &,
                                                            const std::vector<Correspondence::Fixel> &s,
                                                            const std::vector<Correspondence::Fixel> &t) const {
  std::vector<std::vector<Mapping::Entry>> result;
  result.reserve(t.size());
  for (index_type it = 0; it != t.size(); ++it) {
    index_type closest_index = 0;
    float max_dp = 0.0f;
    for (index_type is = 0; is != s.size(); ++is) {
      const float dp = t[it].absdot(s[is]);
      if (dp > max_dp) {
        max_dp = dp;
        closest_index = is;
      }
    }
    if (max_dp > dp_threshold)
      result.push_back({Mapping::Entry{closest_index, 1.0f}});
    else
      result.emplace_back();
  }
  return result;
}

} // namespace MR::Fixel::Correspondence::Algorithms

File: python/mrtrix3/commands/dwi2mask/legacy.py
# Copyright (c) 2008-2026 the MRtrix3 contributors.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Covered Software is provided under this License on an "as is"
# basis, without warranty of any kind, either expressed, implied, or
# statutory, including, without limitation, warranties that the
# Covered Software is free of defects, merchantable, fit for a
# particular purpose or non-infringing.
# See the Mozilla Public License v. 2.0 for more details.
#
# For more details, see http://www.mrtrix.org/.

from mrtrix3 import app, run

NEEDS_MEAN_BZERO = False # pylint: disable=unused-variable
DEFAULT_CLEAN_SCALE = 2


def usage(base_parser, subcommands): #pylint: disable=unused-variable
  parser = subcommands.add_subcommand('legacy', parent=base_parser)
  parser.set_author('Robert E. Smith (robert.smith@florey.edu.au)')
  parser.set_synopsis('Use the legacy MRtrix3 dwi2mask heuristic (based on thresholded trace images)')
  parser.add_argument('input',
                      'The input DWI series',
                      type=app.Parser.ImageIn())
  parser.add_argument('output',
                      'The output mask image',
                      type=app.Parser.ImageOut())
  parser.add_option('clean_scale',
                    'the maximum scale used to cut bridges. '
                    'A certain maximum scale cuts bridges up to a width (in voxels) of 2x the provided scale. '
                    'Setting this to 0 disables the mask cleaning step.',
                    type=app.Parser.Int(0),
                    default=DEFAULT_CLEAN_SCALE)



def execute(): #pylint: disable=unused-variable

  run.command('mrcalc input.mif 0 -max - | '
              'dwishellmath - mean - | '
              'mrthreshold - - -comparison gt | '
              'mrmath - max -axis 3 - | '
              'maskfilter - median - | '
              'maskfilter - connect -largest - | '
              'maskfilter - fill - | '
              f'maskfilter - clean -scale {app.ARGS.clean_scale} mask.mif')

  return 'mask.mif'
